celesta-sql/src/main/java/ru/curs/celesta/score/TableBuilder.java

```diff
@@ -0,0 +1,203 @@
+package ru.curs.celesta.score;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Table builder class that is used by generated {@link ru.curs.celesta.score.CelestaParser} as
+ * a workaround to overcome some shortcomings of JavaCC.
+ *
+ * @author Pavel Perminov (packpaul@mail.ru)
+ * @since 2019-07-16
+ */
+final class TableBuilder {
+
+    private final GrainPart grainPart;
+    private final String name;
+
+    private List<BuildAction> actions = new ArrayList<>();
+
+    private BasicTable table;
+    private boolean isReadOnly;
+
+    TableBuilder(BasicTable table) {
+        this(table.getGrainPart(), table.getName());
+        this.table = table;
+    }
+
+    TableBuilder(GrainPart grainPart, String name) {
+        this.grainPart = grainPart;
+        this.name = name;
+    }
+
+    public Grain getGrain() {
+        return this.grainPart.getGrain();
+    }
+
+    public BasicTable build() throws ParseException {
+
+        if (this.table == null) {
+            this.table = isReadOnly ? new ReadOnlyTable(grainPart, name) : new Table(grainPart, name);
+        }
+
+        for (BuildAction action : actions) {
+            action.execute();
+        }
+
+        this.table.finalizePK();
+
+        return this.table;
+    }
+
+    public void setCelestaDocLexem(String celestaDoc) {
```

Comment:
Just a thought (not insisting): maybe `withCelestaDocLexem`, `withVersioned` etc might be more orthodox naming for builder methods? (I know renaming them implies changing the .jj spec, so, really, I'm not insisting).

Reply:
During refactoring I had to keep the API as the same as of Table, that's why the naming was preserved. Plus TableBuilder is a package private technical class that's used only in CelestaParser, and if needed it can be used interchangeably instead of Table or ReadOnlyTable.